R/unreachable_code_linter.R

```diff
@@ -0,0 +1,45 @@
+#' Block unreachable code and comments following return statements
+#'
+#' Code after a top-level [return()] or [stop()] can't be reached; typically
+#'   this is vestigial code left after refactoring or sandboxing code, which is
+#'   fine for exploration, but shouldn't ultimately be checked in. Comments
+#'   meant for posterity should be placed *before* the final `return()`.
+#'
+#' @evalRd rd_tags("unreachable_code_linter")
+#' @seealso [linters] for a complete list of linters available in lintr.
+#' @export
+unreachable_code_linter <- function() {
+  Linter(function(source_file) {
+    if (length(source_file$xml_parsed_content) == 0L) {
+      return(list())
+    }
+
+    xml <- source_file$xml_parsed_content
+
+    # NB:
+    #  - * returns all children, including the terminal }, so the position
+    #    is not last(), but last()-1. If there's no }, this linter doesn't apply.
+    #  - land on the culprit expression
+    xpath <- "
+    //FUNCTION
+    /following-sibling::expr
+    /*[
+      self::expr
+      and expr[SYMBOL_FUNCTION_CALL[text() = 'return' or text() = 'stop']]
+      and (position() != last() - 1 or not(following-sibling::OP-RIGHT-BRACE))
```

Comment:
Curious: Does this cope properly with return in switch?

Reply:
Indeed it doesn't -- just added an example test (commented out) to make sure I captured what you had in mind.

I think `tryCatch()` is equally affected.